pagopa/pn-frontend
# PN-7438 - cra to vite

## Short description
Adapted the structure of the repo and the projects contained in it, to eliminate the tools and settings provided by Create React App, react-scripts and craco, and use Vite instead. The monorepo still uses Lerna.   
For test execution we still rely on Jest; we leave the replacement with Vitest for a subsequent PR.

## List of changes proposed in this pull request
(there are a lot!) 

Added the files `vite.config.ts` and `vite-env.d.ts` in each package.

In each package, moved the `index.html` from `/public` to the root folder, as required by the standard settings of Vite, and added the line required for Vite to actually load the contents. Also renamed `src/index.tsx` as `src/main.tsx`.

Changed the tasks `start` (which is also renamed as `dev`) to just launch `vite`, and `build` to execute `vite build`.

Added the `preview` task to run the build in locale, is just `vite preview`; cfr. https://vitejs.dev/guide/static-deploy.html#testing-the-app-locally.

Removed the `eject` tasks.

Removed the references to craco and react-scripts in all package.json files, and also all craco-related config files.

Make the versions of the dependencies used in several packages uniform. This helps to avoid weird errors that derive from having many different versions of a single package in the common `yarn.lock`.

Changed the name of env variables from `REACT_APP_XXX` to `VITE_APP_XXX` or just `VITE_XXX`. Also changed some references to `process.env` using `import.meta.env` instead, check https://vitejs.dev/guide/env-and-mode.html.

Made some modifications in the `tsconfig.json` files. In particular, changed the `outDir` to `dist`, this folder were already in the `.gitignore` files.

Change of the path for the `config/config.json` file when fetching it for the configuration service, now it takes into account the `import.meta.env.BASE_URL` variable.  
Check https://vitejs.dev/guide/build.html#public-base-path.

Added the files and dependencies needed to execute test using Jest in absence of the settings provided by react-scripts.
This includes: files `babel.config.js` and `jest.config.js` in each package, file `jest.config.base.js` at the root, severals devDependencies added or updated, all `test:XXX` scripts modified.

Added the option `test` as a possible value for NODE_ENV, in the `react-app-env.d.ts` file for each package.

Added a file `svgTransformerForTests.js` in order to allow the tests involving the load of an SVG file to be charged; this file is referred to in the base Jest config. 
Check https://stackoverflow.com/questions/58603201/jest-cannot-load-svg-file.

Added `await` on the invocations to functions provided by `@testing-library/user-event`.  
Motivation for this: the general upgrade on dependencies includes the upgrade of that package from version 13 to version 14, which implies that now the methods of `userEvent` are async; check https://github.com/testing-library/user-event/releases/tag/v14.0.0.


## How to test
Just enter and navigate each app (this can be done as before through `yarn start:<pf/pg/pa/login>` from the root folder).

For each package, run `yarn build` to verify that no compilation errors arise, and `yarn test` to verify that Jest runs OK.
